src/packages/auth/src/authentication/providers/local/plugin.ts

```diff
@@ -0,0 +1,97 @@
+import { ApolloServerPlugin } from '@apollo/server';
+import { logger } from '@exogee/logger';
+
+import { AuthorizationContext } from '../../../types';
+import { LocalAuthTokenProvider } from './provider';
+import { upsertAuthorizationContext } from '../../../helper-functions';
+import { UserProfile } from '../../../user-profile';
+import { ForbiddenError } from 'apollo-server-errors';
+import { AuthProvider } from '../../base-auth-token-provider';
+import { GraphQLError } from 'graphql';
+
+if (!process.env.LOCAL_AUTH_REDIRECT_URI)
+	throw new Error('LOCAL_AUTH_REDIRECT_URI is required in environment');
+
+const redirectUrl = process.env.LOCAL_AUTH_REDIRECT_URI;
+
+const didEncounterForbiddenError = (error: any) => {
+	return error.extensions.code === 'FORBIDDEN';
+};
+
+export const localAuthApolloPlugin = (
+	addUserToContext: (userId: string) => Promise<UserProfile>
+): ApolloServerPlugin<AuthorizationContext> => ({
+	async requestDidStart({ request, contextValue }) {
+		logger.trace('LocalAuthApolloPlugin requestDidStart');
+		// We have two use cases this needs to handle:
+		// 1. No auth header, initial request for a guest operation.
+		//    - Some requests are allowed when accessed by a guest defined above.
+		// 2. There is an auth header
+		//    - In this situation we need to verify the token, error if it's not valid and redirect.
+		//
+
+		// We may need to return a redirect to the client. If so, we'll set this variable.
+		const authHeader = request.http?.headers.get('authorization');
+		// If verification fails then set this flag
+		let tokenVerificationFailed = false;
+
+		// Case 1. No auth header, initial request for a guest operation.
+		if (!authHeader) {
+			logger.trace('No Auth header, setting redirect');
+
+			// We are a guest and have not logged in yet.
+			contextValue.user = new UserProfile({
+				id: undefined,
+				roles: ['GUEST'],
+				provider: AuthProvider.LOCAL,
+			});
+			upsertAuthorizationContext(contextValue);
+		} else {
+			// Case 2. There is an auth header
+			logger.trace('Got a token, checking it is valid.');
+
+			const tokenProvider = new LocalAuthTokenProvider();
+
+			try {
+				const decoded = await tokenProvider.decodeToken(authHeader);
```

Comment:
We should send the Authorization header as `[mode] [credential]` ie `Bearer (token)` to keep to spec -  should we split that here or in the token provider?

Reply:
yeah ill add this

Reply:
all done